### ISSUE TITLE
Stop-Computer, Restart-Computer: Add support for PowerShell Remoting 

### ISSUE THREAD

**Issue by mh6543** [PowerShell/PowerShell#21180]
### Summary of the new feature / enhancement

As a user, I want to be able to use `Restart-Computer -Wait` on machines that only support PowerShell remoting over SSH, or have a non-default WinRM port for example.

Currently, `Restart-Computer` only uses default DCOM/WSMan/WinRM, and fails to connect in those circumstances

### Proposed technical implementation details (optional)

Proposed: allow using `-Session` parameter from `New-PSSession` or `Invoke-Command` in place of `-ComputerName`:
```
$mySession = New-PSSession -ComputerName "Remote"
Restart-Computer -Wait -Session $mySession
```
Note: by default, a PSSession will go into `State: Broken` when the remote host reboots, which does not allow reconnection. <s>`Restart-Computer` may need to use `Disconnect-PSSession` in the background just *after* issuing the restart, and attempt `Connect-PSSession` to maintain the same session object.</s>

**Comment by rhubarb-geek-nz** [comment]
I don't think the Disconnect-PSSession and Connect-PSSession will work with SSH. If they did they would be available on Linux.

New-PSSession works by creating a live session with sshd/pwsh, you can see it on a server with

```
$ ps -ef | grep pwsh
bythesea   15091 15090 34 18:47 ?        00:00:06 /usr/bin/pwsh -sshs -nologo
```

When the session is closed the SSH connection is closed and the sshd/pwsh pair exit. With a reboot the SSH session will be gone, there is nothing to resume. 

You will need to go through the complete SSH authentication again.

In other words, there is no session to resume once you have ended the SSH/TCP connection, so if you are rebooting a computer you will need to create a new session once it is back up and responding to requests, something you will only be able to do with the client polling ( with timeouts) to attempt to make connections.

**Comment by mh6543** [comment]
I think it's acceptable to create a new PSSession object, which can be done using the existing session:

```
$cred = Get-Credential  ## some other user
$mySession = New-PSSession -ComputerName "Remote" -Credential $cred

Restart-Computer "Remote"
# session broken

## reconnection succeeds
$after = New-PSSession -Session $mySession  
Get-PSSession

 Id Name            Transport ComputerName    ComputerType    State         ConfigurationName     Availability
 -- ----            --------- ------------    ------------    -----         -----------------     ------------
  1 Runspace1       WSMan     Remote          RemoteMachine   Broken        Microsoft.PowerShell          None
  2 Runspace2       WSMan     Remote          RemoteMachine   Opened        Microsoft.PowerShell     Available
```

This probably works for SSH `-KeyFilePath` as well? Interactive password authentication should be fine since we're already waiting.

Then need to output the new session

**Comment by SteveL-MSFT** [comment]
I don't think we should add remoting capabilities to cmdlets and instead prefer cmdlets used over remoting such as `Invoke-Command`.  In the specific case of polling if a computer is back up, it may make sense to have a `Wait-PSSession` cmdlet that waits until it can determine the remote system is up.

**Comment by rhubarb-geek-nz** [comment]
> `Wait-PSSession` cmdlet that waits until it can determine the remote system is up.

Given we need to pass credentials to confirm the system is up and ready and want to create a session, perhaps the wait option should be on the  New-PSSession where instead of a timeout it is given a cancellation token and attempts to make connections until the cancellation token is signalled. (a timeout is just one of the cancellation token mechanisms)

**Comment by mklement0** [comment]
@SteveL-MSFT:

>  don't think we should add remoting capabilities to cmdlets 

While that is true in general, `Restart-Computer` constitutes a special use case and already _does_ have remoting capabilities, albeit currently _on Windows only_.

  * Restarting invariably involves broken remote connections, which would be awkward to handle individually, and so would having to explicitly wait for machines to come back online (a `Wait-PSSession` cmdlet may still make sense _separately_).

  * The existing - currently Windows-only - `-Wait` switch takes the drudgery out of this: it synchronously waits for all targeted remote machines to come back online after a reboot.

This selective, remoting transport-specific / platform-specific support for awaiting remote reboots to complete amounts to an awkward asymmetry, which @mh6543's suggestion aims to overcome.

**Comment by SteveL-MSFT** [comment]
> > `Wait-PSSession` cmdlet that waits until it can determine the remote system is up.
> 
> Given we need to pass credentials to confirm the system is up and ready and want to create a session, perhaps the wait option should be on the New-PSSession where instead of a timeout it is given a cancellation token and attempts to make connections until the cancellation token is signalled. (a timeout is just one of the cancellation token mechanisms)

In the case of WS-Man, it was designed with a single anonymous SOAP endpoint that returns the WS-Man stack version and that's what is used to determine if the system is up.

**Comment by TobiasPSP** [comment]
`-Wait`  on Windows historically probably just was a "low hanging fruit": WsMan provides an anonymous way of getting a response from the restarting system. That's almost like ping/open ICMP with all its implications.

This low hanging fruit does not exist elsewhere (for good).

Only circumstance, individual use-case, and infrastructure can answer the question whether `-Wait`ish behavior can be implemented, and *how*. There's no catch-all logic that could be safely added to `Restart-Process`.

Rather, this is something that needs to be decided and done by the script author: in one case, a simple ping or port test may do, in other cases you might have to actively try and establish new ssh authenticated sessions until they succeed, and yet in other scenarios it's impossible or not desirable or legal to know whether a remote system is up or not.

I'd suggest to not try and port `-Wait` from Windows to other OS.

**Comment by mklement0** [comment]
@TobiasPSP, I can appreciate the technical challenges on reconnection, but what about @mh6543's suggestion to "refresh" the original session with `$after = New-PSSession -Session $mySession`, with `$mySession` representing the session object used initially, via a hypothetical `-Session` argument passed to `Restart-Computer`?

Does this not work in a transport-agnostic fashion? (I'm asking innocently.)

As for scenarios where it's impossible / not desirable or legal to check whether a restart completed: it would then be at the discretion of the user whether to use `-Wait` or not.

**Comment by TobiasPSP** [comment]
d‘accord, if there is a simple way to refresh the session and can be implemented in a OS agnostic way. The moment effort or complexity starts to get out of hand I‘d skip it.

 -Wait in Restart-Computer has a very unique history. Cmdlets generally have no -Wait and expect the user to implement checks if needed, even more so when checks involve remoting. 
 
But maybe I am underestimating the practical use of -Wait, too. Is it in high demand?

**Comment by SteveL-MSFT** [comment]
@PowerShell/wg-powershell-cmdlets reviewed this.  As noted, WSMan has a special anonymous endpoint that allows checking if the system is responding to requests (and thus back up).  Such a mechanism does not exist for SSH.  Using a custom WinRM port means that endpoint can't be discovered and there is no intent to support that capability.
